libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1719,3 +1740,189 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
 	solAssert(_expression.annotation().type, "Type of expression not set.");
 	return *_expression.annotation().type;
 }
+
+bool IRGeneratorForStatements::visit(TryStatement const& _tryStatement)
+{
+	solAssert(!m_returnInfo.has_value(), "");
+
+	m_returnInfo.reset();
+	_tryStatement.externalCall().accept(*this);
+	solAssert(m_returnInfo.has_value(), "");
+
+	auto const trySuccessCondition = m_context.trySuccessConditionVariable(_tryStatement.externalCall());
+
+	m_code << "switch iszero(" << trySuccessCondition << ")\n";
+
+	m_code << "case 0 { // success case\n";
+	TryCatchClause const& successClause = *_tryStatement.clauses().front();
+	decodeTryCallReturnParameters(successClause);
+	m_returnInfo.reset();
+	successClause.block().accept(*this);
+	m_code << "}\n";
+
+	m_code << "default { // failure case\n";
+	handleCatch(_tryStatement);
+	m_code << "}\n";
+
+	return false;
+}
+
+string IRGeneratorForStatements::decodeReturnParameters()
+{
+	ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
+
+	return Whiskers(R"(
+		<?dynamicReturnSize>
+			// copy dynamic return data out
+			returndatacopy(<pos>, 0, returndatasize())
+		</dynamicReturnSize>
+
+		// update freeMemoryPointer according to dynamic return size
+		mstore(<freeMemoryPointer>, add(<pos>, and(add(<returnSize>, 0x1f), not(0x1f))))
+
+		// decode return parameters from external try-call into <retVars>
+		let <retVars> := <abiDecode>(<pos>, add(<pos>, <returnSize>))
+	)")
+	("retVars", IRVariable(m_returnInfo->functionCall).commaSeparatedList())
+	("pos", m_returnInfo->returndataVariable)
+	("abiDecode", abi.tupleDecoder(m_returnInfo->returnTypes, true))
+	("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+	("dynamicReturnSize", m_returnInfo->dynamicReturnSize)
+	("returnSize",
+		m_context.evmVersion().supportsReturndata()
+			? "returndatasize()"s
+			: to_string(m_returnInfo->estimatedReturnSize)
+	).render();
+}
+
+void IRGeneratorForStatements::decodeTryCallReturnParameters(TryCatchClause const& _successClause)
+{
+	solAssert(m_returnInfo.has_value(), "");
+	if (!_successClause.parameters())
+		return;
+
+	m_code << decodeReturnParameters();
+
+	if (_successClause.parameters()->parameters().size() > 1)
+	{
+		size_t i = 0;
+		for (ASTPointer<VariableDeclaration> const& varDecl: _successClause.parameters()->parameters())
+		{
+			solAssert(varDecl, "");
+			define(
+				m_context.addLocalVariable(*varDecl),
+				IRVariable(m_returnInfo->functionCall).tupleComponent(i++)
+			);
+		}
+	}
+	else
+	{
+		define(
+			m_context.addLocalVariable(*_successClause.parameters()->parameters().front()),
+			IRVariable(m_returnInfo->functionCall)
+		);
+	}
+}
+
+void IRGeneratorForStatements::handleCatch(TryStatement const& _tryStatement)
+{
+	if (auto structured = _tryStatement.structuredClause(); structured)
```

Comment:
I think it is better readable to just call the function twice instead of introducing a new variable and using that variable three times.